Fix support for `maxLevelCell = 0` in MPAS-Ocean

xylar opened the pull request:
There are various locations where `maxLevelCell` was previously assumed to be non-zero that this PR fixes.

xylar committed:
Fix zMid, zTop and ssh when maxLevelCell == 0

xylar committed:
Fix eddy parameterization when maxLevelCell == 0

xylar committed:
Fix mono tracer advection when maxLevelCell == 0

xylar committed:
Fix cvmix when maxLevelCell == 0

xylar committed:
Fix vmix when maxLevelCell == 0

xylar committed:
Fix forcing when maxLevelCell == 0

cbegeman committed:
Make MLD AM compatible with maxLevelCell=0

cbegeman committed:
Make surface layer computation compatible with maxLevelCell = 0

cbegeman committed:
Make Redi compatible with maxLevelCell = 0

cbegeman committed:
Make Redi compatible with maxLevelCell = 0 part 2

cbegeman committed:
Make tracer nonlocal flux compatible with maxLevelCell = 0

cbegeman committed:
Make cvmix compatible with maxLevelCell = 0

cbegeman commented:
@xylar I haven't had a chance to look at the output, but these changes resulted in QUwISC240's ssh_adjustment completing when ISC cells have `maxLevelCell = 0`.

xylar commented:
@cbegeman, I noticed that I started from a really old base here (6 months old). Not sure quite why but I was working from Polaris so maybe that was the current state of the E3SM-Project submodule where I was working.  Anyway, I rebased and removed the one commit that wasn't necessary anymore (and had a conflict).

xylar commented:
I'll do some more testing as soon as I have time.  I would like to make a version of the IcoswISC30E3r5 mesh with `maxLevelCell = 0` as a starting point, assuming no one has done that already.

cbegeman commented:
@xylar Yes, I noticed that it was old but didn't want to change your branch too much. Thanks for rebasing!

cbegeman commented:
@xylar Have you checked if this is BFB lately? I could do this. Since we are moving on from this approach, it could be nice to get these fixes in so that it's possible for folks to build and run with `maxLevelCell=0` in the future.

xylar commented:
@cbegeman, no, I haven't done any BFB testing of this but I would be very surprised if it isn't BFB on meshes that don't have `maxLevelCell = 0`.

I would still like to test it more on the ISOMIP+ test cases in Polaris, though, before I try to get it merged into E3SM just so I don't end up needing to do a second PR for things I missed.  I think I might be able to get to that tomorrow or before early next week at the latest.

cbegeman commented:
@xylar That makes sense. Thanks! I wasn't sure if you were still going this route for ISOMIP+ and I wanted to make sure it didn't get stranded.
